src/modules/mavlink/streams/FUEL_STATUS.hpp

```diff
@@ -0,0 +1,86 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2024 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#ifndef FUEL_STATUS_HPP
+#define FUEL_STATUS_HPP
+
+#include <uORB/topics/fuel_tank_status.h>
+
+class MavlinkStreamFuelStatus : public MavlinkStream
+{
+public:
+	static MavlinkStream *new_instance(Mavlink *mavlink) { return new MavlinkStreamFuelStatus(mavlink); }
+
+	static constexpr const char *get_name_static() { return "FUEL_STATUS"; }
+	static constexpr uint16_t get_id_static() { return MAVLINK_MSG_ID_FUEL_STATUS; }
+
+	const char *get_name() const override { return MavlinkStreamFuelStatus::get_name_static(); }
+	uint16_t get_id() override { return get_id_static(); }
+
+	unsigned get_size() override
+	{
+		return _fuel_tank_status_sub.advertised() ? MAVLINK_MSG_ID_FUEL_STATUS_LEN + MAVLINK_NUM_NON_PAYLOAD_BYTES : 0;
+	}
+
+private:
+	explicit MavlinkStreamFuelStatus(Mavlink *mavlink) : MavlinkStream(mavlink) {}
+
+	uORB::Subscription _fuel_tank_status_sub{ORB_ID(fuel_tank_status)};
+
+	bool send() override
+	{
+		fuel_tank_status_s fuel_status;
+
+		if (_fuel_tank_status_sub.update(&fuel_status)) {
+			mavlink_fuel_status_t msg{};
+
+			msg.id = fuel_status.fuel_tank_id;
+			msg.maximum_fuel = fuel_status.maximum_fuel_capacity;
```

Comment:
Do all of these values set the invalid value - NaN / UINT8 or whatever when the information is not supplied?

The source does not in theory have to be DroneCAN

Reply:
As per the [DroneCAN message](https://github.com/dronecan/DSDL/blob/master/uavcan/equipment/ice/1129.FuelTankStatus.uavcan#L3):

```
All fields are required unless stated otherwise. Unpopulated optional fields should be set to NaN.
```

So it's expected that the DroneCAN message sets this fields. The only field that is optional in the `FuelTankStatus` message is the temperature field, to which we are adding that same check (and again, would expect that the value would be NaN when coming from DroneCAN, we are just adding an extra check).

Reply:
My point was that it might be worth having the checks on what is in the uorb in the mavlink streaming code rather than assuming the uorb topics are populated as per MAVLink expects. That way if a fuel tank is populated from some other source in future things don't break.

Perhaps documenting the required invalid value in the uorb topic makes sense.

I'm not a programmer, so if there is no risk of need for this kind of thinking, ignore.

Reply:
good points. done here: https://github.com/PX4/PX4-Autopilot/pull/23137/commits/4249c5cca13b10d6423b1ef657aaf22cc5fa1d35